client/consensus/grandpa/authorities.go

```diff
@@ -0,0 +1,908 @@
+	inner setIDNumber[N]
+}
+
+func (set[N]) Index() uint {
+	return 1
+}
+
+type unknown struct{}
+
+func (unknown) Index() uint {
+	return 2
+}
+
+// Three states that can be returned: Latest, Set (tuple), Unknown
+func (asc *AuthoritySetChanges[N]) getSetID(blockNumber N) (authSetChangeID authoritySetChangeID, err error) {
+	if asc == nil {
+		return authSetChangeID, fmt.Errorf("getSetID: authSetChanges is nil")
+	}
+	authSetChangeID = newAuthoritySetChangeID[N]()
+	authSet := *asc
+	last := authSet[len(authSet)-1]
+	if last.blockNumber < blockNumber {
+		err = authSetChangeID.Set(latest{})
+		if err != nil {
+			return authSetChangeID, err
+		}
+		return authSetChangeID, nil
+	}
+
+	idx, _ := slices.BinarySearchFunc(
+		authSet,
+		blockNumber,
+		func(a setIDNumber[N], b N) int {
+			switch {
+			case a.blockNumber == b:
+				return 0
+			case a.blockNumber > b:
+				return 1
+			case a.blockNumber < b:
+				return -1
+			default:
+				panic("huh?")
+			}
+		},
+	)
+	if idx < len(authSet) {
+		authChange := authSet[idx]
+
+		// if this is the first index but not the first set id then we are missing data.
+		if idx == 0 && authChange.setID != 0 {
+			err = authSetChangeID.Set(unknown{})
+			if err != nil {
+				return authSetChangeID, err
+			}
+			return authSetChangeID, nil
+		}
+		err = authSetChangeID.Set(set[N]{
+			authChange,
+		})
+		if err != nil {
+			return authSetChangeID, err
+		}
+		return authSetChangeID, nil
+	}
+
+	err = authSetChangeID.Set(unknown{})
+	if err != nil {
+		return authSetChangeID, err
+	}
+	return authSetChangeID, nil
+}
+
+func (asc *AuthoritySetChanges[N]) insert(blockNumber N) {
+	var idx int
+	if asc == nil {
+		panic("authority set changes must be initialised")
+	} else {
+		idx, _ = slices.BinarySearchFunc(
+			*asc,
+			blockNumber,
+			func(a setIDNumber[N], b N) int {
+				switch {
+				case a.blockNumber == b:
+					return 0
+				case a.blockNumber > b:
+					return 1
+				case a.blockNumber < b:
+					return -1
+				default:
+					panic("huh?")
+				}
+			},
+		)
+	}
+
+	set := *asc
+
+	var setId uint64
+	if idx == 0 {
+		setId = 0
+	} else {
+		setId = set[idx-1].setID + 1
+	}
+
+	if idx != len(set) && set[idx].setID == setId {
+		panic("inserting authority set hashNumber")
+	}
+
+	change := setIDNumber[N]{
+		setID:       setId,
+		blockNumber: blockNumber,
+	}
+
+	// Insert hashNumber at index
+	if len(set) <= idx {
+		set = append(set, change)
+	} else {
+		set = append(set[:idx+1], set[idx:]...)
+		set[idx] = change
+	}
+	*asc = set
+}
+
+// IterFrom This logic is used in warp sync proof
+func (asc *AuthoritySetChanges[N]) IterFrom(blockNumber N) *AuthoritySetChanges[N] {
+	if asc == nil {
+		return nil
+	}
+	authSet := *asc
+
+	idx, found := slices.BinarySearchFunc(
+		*asc,
+		blockNumber,
+		func(a setIDNumber[N], b N) int {
+			switch {
+			case a.blockNumber == b:
+				return 0
+			case a.blockNumber > b:
+				return 1
+			case a.blockNumber < b:
+				return -1
+			default:
+				panic("huh?")
```

Comment:
looks like all possible comparisons are covered so I think you can remove the `default` case
```suggestion
```

Reply:
and here